First pull request on level-3 branch. Front-end part - all the functionality of the app (figma: dark-4)

letStayFoolish opened the pull request:
I look forward to your feedback and collaboration as we continue to refine and evolve this project.

letStayFoolish committed:
first commit

letStayFoolish committed:
Add: Navbar, Header (js + css)

letStayFoolish committed:
Add: Promo component (js + css)

letStayFoolish committed:
Add: Techs component (js + css)

letStayFoolish committed:
Update: Reposition between Header and Navbar, so Navbar would always be top over the Header component

letStayFoolish committed:
Update: Main Page done

Finish Components:
- JSX ok!
- CSS ok!
- Responsive ok!

letStayFoolish committed:
Update: Header, Navigation, Promo

Add Navigation component

Correction: Header(old) -> Promo(new) & Navbar(old) -> Header(new)

letStayFoolish committed:
Add: Burger-menu

Add animations
add burger-menu on screens < 769px

letStayFoolish committed:
Add: Movies & MoviesCard components

JSX + CSS

letStayFoolish committed:
Add: Movies, MoviesCard and SearchForm Responsive (CSS)

letStayFoolish committed:
Add: React Routing

Created routes for components:
- Main (home page)
- Movies
- Saved Movies

letStayFoolish committed:
Update: Compress all of PNG and SVG images

letStayFoolish committed:
Update: Navigation component

Account icon background color updates depending on path location (page URL)

letStayFoolish committed:
Add: NotFound Component

Created JSX and CSS + responsive component

letStayFoolish committed:
Add: Signup & Signin components

- JSX
- CSS3
- Responsive

letStayFoolish committed:
Add: Profile component

correct: Responsiveness
create: global components: Menu, Entry (signin + signup)

letStayFoolish committed:
Update: correction with toggleMenu. Added toggleMenu as a props to Menu component on small screens.

letStayFoolish committed:
Update: corrected Menu width on 768px and less

letStayFoolish committed:
Update: corrected SavedMovies component grid columns width on various device screens

letStayFoolish committed:
Update: Main page responsive corrections

letStayFoolish committed:
Update: NotFound component responsive corrections

letStayFoolish committed:
Update: Movies

Correction with media queries

letStayFoolish committed:
Update: Movies

Correction with media queries

letStayFoolish committed:
Update: Readme fil

letStayFoolish committed:
Update: Readme fil

letStayFoolish committed:
Update: Navigation component

Added overlay container on screen width 768px (only)

letStayFoolish committed:
Update: alt text changes in Menu.jsx

letStayFoolish committed:
Update: Input

Added placeholders on input fields.

letStayFoolish committed:
Fix: Revision #1 (after first code review)

- Header: position: from fixed to absolute
- Navbar: menu background color change to dark on every page.
- Close button imported new icon and change positions on max-width: 768px, screen devices
- Menu: font size change to 18px on max-width: 768px
- Profile: because header needs to be !fixed added to profile__heading margin-top
- About Me: GitHub link - added opacity on hover & target="_blank"
- Project: list of project within <ul><li>...</li></ul>. h2 and img block wrapped within <a></a> link element.
- Signup & Signin: click on logo -> '/' page
- index.css: Add alternative font - Arial
- Fix: remove unnecessary <div> elements
- Update: Burger menu icon, exported from design
- Fix: Context centering horizontal
- Fix: Account div background
-

letStayFoolish committed:
Update: Clear code from unused elements

letStayFoolish committed:
Update: Add required on Input elements

letStayFoolish committed:
Update: Convert rem to px

letStayFoolish committed:
Level-3 | Add: Register logic

letStayFoolish committed:
Level-3 | Add: Login and checkToken

letStayFoolish committed:
Level-3 | Add: Logout

Also update Login and Register logic within component

letStayFoolish committed:
Level-3 | Add: SearchForm logic

- Render only movies that match text in search form (typed by user)
- Render Preloader component while waiting for movies to be shown.

letStayFoolish committed:
Level-3 | Add: Movies logic

- first search session: request get all movies from the server
- show filtred movies (short or not)
- save search session to local storage
- if there is data in local storage search in there for movies, checkBox state, if not request from server all movies and then search and filter in there

letStayFoolish committed:
Level-3 | Update: NotFound

- navigate to previous page

letStayFoolish committed:
Level-3 | Add: handler remove saved movies

letStayFoolish committed:
Level-3 | update: Remove trash (commented code lines)

letStayFoolish committed:
Level-3 | update: Save/delete movie to the saved movies list

letStayFoolish committed:
Level-3 | update: Save/delete movie to the saved movies list

letStayFoolish committed:
Level-3 | Upadate: Edit profile

Validation is updated so user can not update provile information if input fields are not filled in correct

letStayFoolish committed:
Level-3 | Update: add real domain

Deployed backend and frontend, for it to work, I have changed BASE_URL in constants to: https://api.more-movies.nomoredomainsicu.ru

letStayFoolish committed:
Level-3 | Fix: Register and Login elements are protected routes also now

Register and Login components wrapped in a protected route element, so user that is already loggedin can't go to the Login/Register page while he is loggedin.

letStayFoolish committed:
Level-3 | Update: BASE_URL changed for site to work on server

to fix CORS problem

letStayFoolish committed:
Level-3 | Update: remove unnecessary code lines

letStayFoolish committed:
Level-3 | FIX: BASE_URL

letStayFoolish committed:
Level-3 | Ready to deploy and code review

letStayFoolish committed:
Update README.md

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация», если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
    - Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
**Комментарий: после успешной регистрации должен автоматически происходить вход и редирект на `/movies`**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
**Комментарий: после успешной авторизации должен автоматически происходить редирект на `/movies`**
    - Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    - Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
      - Поиск фильмов регистронезависимый.
      - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
      - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
    - Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка «Ещё» скрывается.
**Комментарий: https://skr.sh/sL7G8F0GCav по заданию количество карточек должно зависеть от размера экрана и вашего макета. Сейчас условие по выдаче начального количества карточек при поиске и добавляемое количество карточек при нажатии "Еще" не соответствует требованию задания.**
**Комментарий: https://skr.sh/sL7G8F0GCav Необходимо ориентироваться на фактическое изменение количества карточек в строке. Если карточки уже выдаются по две в ряду, значит при любом поиске должно быть выдано 8 и кнопка "Еще" должна добавлять 2. Проверьте все точки переходов и соответствие выдаваемых результатов условию задания. Например, на разрешении 887px фильмы выводятся по 2 в ряду, однако кнопка "Еще" добавляет по 3. Так-же проверьте мобильную версию.**
**Комментарий: При выполнении поиска в блок результата необходимо выдавать то количество карточек, которое соответствует условию задания для текущего разрешения экрана. Сейчас выдается то количество, которое уже было выдано до поиска. Например, при размещении карточек в одну колонку поиск должен выдавать 5 карточек. Если нажать кнопку "Еще", то добавится 2 и в блоке результата будет уже 7. Теперь при выполнении нового поиска в блок результата нужно вывести снова 5 карточек, но сейчас будет выведено 7, как и было до поиска.**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
    - Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
**Комментарий: при нажатии на активную кнопку лайка должен отправляться запрос на удаление**
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`,
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [ ] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
**Комментарий: контекст создан, но фактически не используется вне App**
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске, все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке, данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase,
  - имена переменных — существительные,
  - имена переменных, функций и компонентов соответствуют содержимому,
  - имена коллекций NodeList — существительные во множественном числе,
  - имя функции начинается с глагола и отражает то, что она делает,
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки,
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
**Комментарий: не нужно очищать поля формы до ответа сервера подтверждением, у пользователя должна быть возможность исправить данные, а не вводить их заново. На время выполнения запросанужно только блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при поиске или добавляемое при нажатии кнопки "Еще" на разных размерах экрана и точки перестроения экрана нужно вынести в константы**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде,
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика,
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 60

letStayFoolish committed:
Level-3 | Update: Fix remove saved movies from movies page

letStayFoolish committed:
Level-3 | Update: Fix movie list - cards are showing correct

Numbers of cards to be rendered depending on screen width are working correctly

letStayFoolish committed:
Level-3 | Update: Fix movie list - cards are showing correct

Numbers of cards to be rendered depending on screen width are working correctly

letStayFoolish committed:
Level-3 | Update: Fix user can not update profile with uncorrected data

Fixed so user can not edit and update profile if such an email is already registered in the database.

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация», если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
    - Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
    - Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    - Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
      - Поиск фильмов регистронезависимый.
      - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
      - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
    - Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка «Ещё» скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
    - Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [ ] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
**Комментарий: В мобильном варианте, при отображении карточек в одну колонку, кнопка "Еще" должна добавлять 2 карточки, а не одну.**
**Комментарий: При выполнении поиска в блок результата необходимо выдавать то количество карточек, которое соответствует условию задания для текущего разрешения экрана. Сейчас выдается то количество, которое уже было выдано до поиска. Например, при размещении карточек в одну колонку поиск должен выдавать 5 карточек. Если нажать кнопку "Еще", то добавится 2 и в блоке результата будет уже 7. Теперь при выполнении нового поиска в блок результата нужно вывести снова 5 карточек, но сейчас будет выведено 7, как и было до поиска.**
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`,
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске, все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке, данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase,
  - имена переменных — существительные,
  - имена переменных, функций и компонентов соответствуют содержимому,
  - имена коллекций NodeList — существительные во множественном числе,
  - имя функции начинается с глагола и отражает то, что она делает,
  - для именования запрещены: транслит и неуместные сокращения.


### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки,
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде,
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика,
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 90

## Хорошая работа, весь необходимый функционал работает. Поздравляю с завершением дипломного проекта!

==========================================================

Однако, если планируете использовать проект в портфолио, то советую ещё поработать над ним, исправить оставшиеся замечания и ещё раз перепроверить функционал

=================================================================

### Несколько советов для дальнейшего развития:
- сверстайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- еще немного попрактиковаться с React, например изучить redux
- будет большим плюсом при поиске работы, если немного изучите TypeScript т.к. на него переходят многие компании
- ходить на собеседования, опыт прохождения собеседований не менее важен, чем знание теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушайте подкаксты, они очень хорошо расширяют кругозор, и позволяют узнать, что происходит в отрасли:
https://podlodka.io/
https://vk.com/everyonecanyp
https://music.yandex.ru/album/6880234?dir=desc&activeTab=about
https://soundcloud.com/begebot
https://web-standards.ru/podcast/
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

letStayFoolish committed:
Level-3 | Update: Correct number of cards to be rendered on mobile devices

letStayFoolish committed:
Level-3 | Update: Disable input fields

- On Profile page, corrected so inputs stay disabled while waiting response from server

letStayFoolish committed:
Level-3 | Update: Disable input fields

- On Register & Login page, corrected to disable button and inputs fields on waiting response from server

letStayFoolish committed:
Level-3 | Update: Disable input fields

- Ready to publish
